File: sdks/java/src/test/java/dev/crabka/sdk/QueuesTest.java
package dev.crabka.sdk;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.google.protobuf.ByteString;
import crabka.gateway.v1.GatewayOuterClass;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CopyOnWriteArrayList;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Protocol;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okio.Buffer;
import org.junit.jupiter.api.Test;

final class QueuesTest {
    private static final MediaType CONNECT_PROTO = MediaType.get("application/proto");

    @Test
    void mockAcquireAcknowledgeRenewAndReleaseUseQueueShapes() {
        CrabkaClient client = CrabkaClient.builder().endpoint("mock://gateway").build();
        client.messaging().publish(new Record("queue", bytes("first"), List.of(new Header("kind", bytes("queue"))))).join();
        client.messaging().publish(Record.of("queue", bytes("second"))).join();

        QueueAcquireResult acquired = client.queues().acquire("queue", "workers", 1, Duration.ofSeconds(30)).join();

        assertEquals("queue-session-1", acquired.sessionId());
        assertEquals(1, acquired.messages().size());
        QueueMessage message = acquired.messages().get(0);
        assertEquals("queue:0:0", message.messageId());
        assertEquals("queue", message.topic());
        assertEquals(0, message.partition());
        assertEquals(0, message.offset());
        assertEquals(1, message.deliveryCount());
        assertArrayEquals(bytes("first"), message.value());
        assertEquals(1, message.headers().size());
        assertEquals("kind", message.headers().get(0).name());
        assertArrayEquals(bytes("queue"), message.headers().get(0).value());

        QueueBatchResult renewed = client.queues().renew(acquired.sessionId(), List.of(
                new QueueRenewEntry("queue:0:0"),
                new QueueRenewEntry("missing:0:0"))).join();
        assertQueueResults(renewed.results(), List.of(
                QueueResult.success("queue:0:0"),
                QueueResult.notAcquired("missing:0:0")));

        QueueBatchResult acknowledged = client.queues().acknowledge(acquired.sessionId(), List.of(
                new QueueAckEntry("queue:0:0", QueueAckType.RELEASE))).join();
        assertQueueResults(acknowledged.results(), List.of(QueueResult.success("queue:0:0")));

        QueueAcquireResult redelivered = client.queues()
                .acquireWithSession("queue", "workers", 1, Duration.ofSeconds(30), acquired.sessionId())
                .join();
        assertEquals(acquired.sessionId(), redelivered.sessionId());
        assertEquals("queue:0:0", redelivered.messages().get(0).messageId());
        assertEquals(2, redelivered.messages().get(0).deliveryCount());
    }

    @Test
    void validationErrorsUseConformanceMessages() {
        CrabkaClient client = CrabkaClient.builder().endpoint("mock://gateway").build();

        assertInvalidArgument("queue group is required",
                () -> client.queues().acquire("queue", "", 1, Duration.ofSeconds(30)).join());
        assertInvalidArgument("queue lock_duration_ms must be 30000; per-acquire lock durations are not supported",
                () -> client.queues().acquire("queue", "workers", 1, Duration.ofSeconds(1)).join());
        assertInvalidArgument("queue session_id is required",
                () -> client.queues().renew("", List.of()).join());
    }

    @Test
    void mockSessionsOwnDeliveredCoordinatesAndRejectUnknownReuse() {
        CrabkaClient client = CrabkaClient.builder().endpoint("mock://gateway").build();
        client.messaging().publish(Record.of("queue", bytes("job"))).join();
        QueueAcquireResult first = client.queues().acquire("queue", "workers", 1, Duration.ofSeconds(30)).join();
        QueueAcquireResult second = client.queues().acquire("queue", "workers", 1, Duration.ofSeconds(30)).join();

        QueueBatchResult acknowledged = client.queues().acknowledge(second.sessionId(), List.of(
                new QueueAckEntry(first.messages().get(0).messageId(), QueueAckType.ACCEPT))).join();
        assertQueueResults(acknowledged.results(), List.of(QueueResult.notAcquired(first.messages().get(0).messageId())));
        QueueBatchResult renewed = client.queues().renew(second.sessionId(), List.of(
                new QueueRenewEntry(first.messages().get(0).messageId()))).join();
        assertQueueResults(renewed.results(), List.of(QueueResult.notAcquired(first.messages().get(0).messageId())));

        assertInvalidArgument("queue session expired; re-acquire", () -> client.queues()
                .acquireWithSession("queue", "workers", 1, Duration.ofSeconds(30), "missing-session")
                .join());
        assertInvalidArgument("group_id and topics are fixed when a queue session is created", () -> client.queues()
                .acquireWithSession("queue", "other-workers", 1, Duration.ofSeconds(30), first.sessionId())
                .join());
    }

    @Test
    void mockQueueStateIsIndependentPerGroup() {
        CrabkaClient client = CrabkaClient.builder().endpoint("mock://gateway").build();
        client.messaging().publish(Record.of("queue", bytes("job"))).join();
        QueueAcquireResult first = client.queues()
                .acquire("queue", "first-workers", 1, Duration.ofSeconds(30))
                .join();
        QueueAcquireResult second = client.queues()
                .acquire("queue", "second-workers", 1, Duration.ofSeconds(30))
                .join();

        assertEquals("queue:0:0", first.messages().get(0).messageId());
        assertEquals(1, first.messages().get(0).deliveryCount());
        assertEquals("queue:0:0", second.messages().get(0).messageId());
        assertEquals(1, second.messages().get(0).deliveryCount());
        assertQueueResults(client.queues().acknowledge(first.sessionId(), List.of(
                new QueueAckEntry("queue:0:0", QueueAckType.RELEASE))).join().results(),
                List.of(QueueResult.success("queue:0:0")));
        assertQueueResults(client.queues().renew(second.sessionId(), List.of(
                new QueueRenewEntry("queue:0:0"))).join().results(),
                List.of(QueueResult.success("queue:0:0")));

        QueueAcquireResult redelivered = client.queues()
                .acquireWithSession("queue", "first-workers", 1, Duration.ofSeconds(30), first.sessionId())
                .join();
        assertEquals(2, redelivered.messages().get(0).deliveryCount());
        assertQueueResults(client.queues().acknowledge(first.sessionId(), List.of(
                new QueueAckEntry("queue:0:0", QueueAckType.ACCEPT))).join().results(),
                List.of(QueueResult.success("queue:0:0")));
        assertQueueResults(client.queues().acknowledge(second.sessionId(), List.of(
                new QueueAckEntry("queue:0:0", QueueAckType.ACCEPT))).join().results(),
                List.of(QueueResult.success("queue:0:0")));
    }

    @Test
    void liveQueueRpcRequestsAndResponsesAreMapped() throws Exception {
        CopyOnWriteArrayList<Request> requests = new CopyOnWriteArrayList<>();
        OkHttpClient httpClient = new OkHttpClient.Builder()
                .addInterceptor(chain -> {
                    Request request = chain.request();
                    requests.add(request);
                    return switch (request.url().encodedPath()) {
                        case "/crabka.gateway.v1.Gateway/QueueAcquire" -> gatewayResponse(queueAcquireResponse());
                        case "/crabka.gateway.v1.Gateway/QueueAcknowledge" -> gatewayResponse(queueBatchResponse(false));
                        case "/crabka.gateway.v1.Gateway/QueueRenew" -> gatewayResponse(queueBatchResponse(true));
                        default -> throw new AssertionError("unexpected gateway path " + request.url().encodedPath());
                    };
                })
                .build();
        LiveGatewayTransport transport = new LiveGatewayTransport(URI.create("http://gateway.test"), "", httpClient);

        QueueAcquireResult acquired = transport.queueAcquire("queue", "workers", 1, 30_000, "actual-session");
        QueueBatchResult acknowledged = transport.queueAcknowledge("actual-session", List.of(
                new QueueAckEntry("queue:0:7", QueueAckType.ACCEPT)));
        QueueBatchResult renewed = transport.queueRenew("actual-session", List.of(
                new QueueRenewEntry("queue:0:7")));

        assertEquals("actual-session", acquired.sessionId());
        assertEquals("queue:0:7", acquired.messages().get(0).messageId());
        assertArrayEquals(bytes("live"), acquired.messages().get(0).value());
        assertQueueResults(acknowledged.results(), List.of(QueueResult.success("queue:0:7")));
        assertQueueResults(renewed.results(), List.of(new QueueResult(
                "queue:0:7",
                new QueueOperationError("invalid_argument", "record is not acquired by this session"))));
        assertEquals(List.of(
                        "/crabka.gateway.v1.Gateway/QueueAcquire",
                        "/crabka.gateway.v1.Gateway/QueueAcknowledge",
                        "/crabka.gateway.v1.Gateway/QueueRenew"),
                requests.stream().map(request -> request.url().encodedPath()).toList());

        GatewayOuterClass.QueueAcquireRequest acquireRequest = GatewayOuterClass.QueueAcquireRequest.parseFrom(bodyBytes(requests.get(0)));
        assertEquals("workers", acquireRequest.getGroupId());
        assertEquals(List.of("queue"), acquireRequest.getTopicsList());
        assertEquals(1, acquireRequest.getMaxMessages());
        assertEquals(30_000, acquireRequest.getLockDurationMs());
        assertEquals("actual-session", acquireRequest.getSessionId());

        GatewayOuterClass.QueueAcknowledgeRequest ackRequest = GatewayOuterClass.QueueAcknowledgeRequest.parseFrom(bodyBytes(requests.get(1)));
        assertEquals("actual-session", ackRequest.getSessionId());
        assertEquals(GatewayOuterClass.QueueAckType.ACCEPT, ackRequest.getEntries(0).getType());

        GatewayOuterClass.QueueRenewRequest renewRequest = GatewayOuterClass.QueueRenewRequest.parseFrom(bodyBytes(requests.get(2)));
        assertEquals("actual-session", renewRequest.getSessionId());
        assertEquals(GatewayOuterClass.QueueAckType.ACCEPT, renewRequest.getEntries(0).getType());
    }

    @Test
    void liveQueueEntryErrorsPreserveGatewayErrorInfo() {
        OkHttpClient httpClient = new OkHttpClient.Builder()
                .addInterceptor(chain -> gatewayResponse(queueBatchErrorResponse(
                        gatewayError(13, "coordinator unavailable", true),
                        gatewayError(13, "commit failed", false),
                        gatewayError(9, "coordinator retry", true))))
                .build();
        LiveGatewayTransport transport = new LiveGatewayTransport(URI.create("http://gateway.test"), "", httpClient);

        QueueBatchResult result = transport.queueAcknowledge("actual-session", List.of(
                new QueueAckEntry("queue:0:7", QueueAckType.ACCEPT),
                new QueueAckEntry("queue:0:8", QueueAckType.RELEASE),
                new QueueAckEntry("queue:0:9", QueueAckType.RELEASE)));

        assertQueueResults(result.results(), List.of(
                new QueueResult("queue:0:9", new QueueOperationError("transport", "coordinator unavailable", true)),
                new QueueResult("queue:0:8", new QueueOperationError("server_error", "commit failed")),
                new QueueResult("queue:0:7", new QueueOperationError("transport", "coordinator retry", true))));
    }

    @Test
    void liveQueueResultsRequireAuthoritativeResponseEntries() {
        OkHttpClient httpClient = new OkHttpClient.Builder()
                .addInterceptor(chain -> gatewayResponse(queueBatchResponseWithoutEntry()))
                .build();
        LiveGatewayTransport transport = new LiveGatewayTransport(URI.create("http://gateway.test"), "", httpClient);

        TransportException error = assertThrows(TransportException.class, () -> transport.queueAcknowledge(
                "actual-session", List.of(new QueueAckEntry("queue:0:7", QueueAckType.ACCEPT))));

        assertEquals("queue response result did not include an entry", error.getMessage());
    }

    private static void assertInvalidArgument(String message, Runnable action) {
        CompletionException error = assertThrows(CompletionException.class, action::run);
        InvalidArgumentException invalidArgument = assertInstanceOf(InvalidArgumentException.class, error.getCause());
        assertEquals(message, invalidArgument.getMessage());
    }

    private static void assertQueueResults(List<QueueResult> actual, List<QueueResult> expected) {
        assertEquals(expected.size(), actual.size());
        for (int index = 0; index < expected.size(); index++) {
            assertEquals(expected.get(index).messageId(), actual.get(index).messageId());
            if (expected.get(index).error() == null) {
                assertNull(actual.get(index).error());
                continue;
            }
            assertEquals(expected.get(index).error(), actual.get(index).error());
        }
    }

    private static byte[] queueAcquireResponse() {
        GatewayOuterClass.QueuedMessage message = GatewayOuterClass.QueuedMessage.newBuilder()
                .setTopic("queue")
                .setPartition(0)
                .setOffset(7)
                .setValue(ByteString.copyFrom(bytes("live")))
                .setDeliveryCount(3)
                .build();
        return GatewayOuterClass.QueueAcquireResponse.newBuilder()
                .setSessionId("actual-session")
                .addMessages(message)
                .build()
                .toByteArray();
    }

    private static byte[] queueBatchResponse(boolean withError) {
        GatewayOuterClass.QueueAckResult.Builder result = GatewayOuterClass.QueueAckResult.newBuilder()
                .setEntry(queueAckEntry(7));
        if (withError) {
            result.setError(GatewayOuterClass.ErrorInfo.newBuilder()
                    .setCode(9)
                    .setMessage("record is not acquired by this session"));
        }
        return GatewayOuterClass.QueueAcknowledgeResponse.newBuilder()
                .addResults(result)
                .build()
                .toByteArray();
    }

    private static byte[] queueBatchErrorResponse(GatewayOuterClass.ErrorInfo... errors) {
        GatewayOuterClass.QueueAcknowledgeResponse.Builder response = GatewayOuterClass.QueueAcknowledgeResponse.newBuilder();
        for (int index = 0; index < errors.length; index++) {
            response.addResults(GatewayOuterClass.QueueAckResult.newBuilder()
                    .setEntry(queueAckEntry(9 - index))
                    .setError(errors[index]));
        }
        return response.build().toByteArray();
    }

    private static byte[] queueBatchResponseWithoutEntry() {
        return GatewayOuterClass.QueueAcknowledgeResponse.newBuilder()
                .addResults(GatewayOuterClass.QueueAckResult.newBuilder())
                .build()
                .toByteArray();
    }

    private static GatewayOuterClass.QueueAckEntry queueAckEntry(long offset) {
        return GatewayOuterClass.QueueAckEntry.newBuilder()
                .setTopic("queue")
                .setPartition(0)
                .setOffset(offset)
                .build();
    }

    private static GatewayOuterClass.ErrorInfo gatewayError(int code, String message, boolean retriable) {
        return GatewayOuterClass.ErrorInfo.newBuilder()
                .setCode(code)
                .setMessage(message)
                .setRetriable(retriable)
                .build();
    }

    private static Response gatewayResponse(byte[] body) {
        return new Response.Builder()
                .request(new Request.Builder().url("http://gateway.test").build())
                .protocol(Protocol.HTTP_2)
                .code(200)
                .message("OK")
                .header("Content-Type", CONNECT_PROTO.toString())
                .body(ResponseBody.create(body, CONNECT_PROTO))
                .build();
    }

    private static byte[] bodyBytes(Request request) throws Exception {
        Buffer buffer = new Buffer();
        request.body().writeTo(buffer);
        return buffer.readByteArray();
    }

    private static byte[] bytes(String value) {
        return value.getBytes(StandardCharsets.UTF_8);
    }
}

File: sdks/ts/src/queues.test.ts
import { describe, expect, test } from "vitest";
import { createClient } from "./client.ts";
import { InvalidArgumentError } from "./errors.ts";
import { createQueuesModule } from "./queues.ts";

const encoder = new TextEncoder();

describe("queues", () => {
  test("mock acquire returns queue v1.1 message shape", async () => {
    const client = createClient({ endpoint: "mock://gateway" });
    await client.messaging.publish({
      topic: "queue-acquire",
      value: encoder.encode("work"),
      headers: [{ name: "kind", value: encoder.encode("queue") }],
    });

    await expect(
      client.queues.acquire("queue-acquire", { group: "workers", max: 1, lockDurationMs: 30_000 }),
    ).resolves.toMatchObject({
      sessionId: "queue-session-1",
      messages: [
        {
          messageId: "queue-acquire:0:0",
          topic: "queue-acquire",
          partition: 0,
          offset: 0n,
          value: encoder.encode("work"),
          headers: [{ name: "kind", value: encoder.encode("queue") }],
          deliveryCount: 1,
        },
      ],
    });
  });

  test("mock acknowledge and renew return per-entry result shapes", async () => {
    const client = createClient({ endpoint: "mock://gateway" });
    await client.messaging.publish({ topic: "queue-ack", value: encoder.encode("ack") });
    const acquired = await client.queues.acquire("queue-ack", { group: "workers", max: 1, lockDurationMs: 30_000 });

    await expect(client.queues.renew(acquired.sessionId, [{ messageId: "queue-ack:0:0" }])).resolves.toEqual({
      results: [{ messageId: "queue-ack:0:0", error: null }],
    });
    await expect(
      client.queues.acknowledge(acquired.sessionId, [
        { messageId: "queue-ack:0:0", ackType: "accept" },
        { messageId: "missing:0:0", ackType: "accept" },
      ]),
    ).resolves.toEqual({
      results: [
        { messageId: "queue-ack:0:0", error: null },
        {
          messageId: "missing:0:0",
          error: { kind: "invalid_argument", message: "queue message is not acquired", retriable: false },
        },
      ],
    });
  });

  test("live queue offsets and message ids preserve bigint precision", async () => {
    const offset = 9_007_199_254_740_993n;
    const gateway = {
      async queueAcquire(request: { sessionId: string }) {
        expect(request.sessionId).toBe("session");
        return { sessionId: "session", messages: [{ topic: "large", partition: 0, offset, value: new Uint8Array(), headers: [], deliveryCount: 1 }] };
      },
      async queueAcknowledge(request: { entries: { offset: bigint }[] }) {
        expect(request.entries[0]?.offset).toBe(offset);
        return { results: [{ entry: { topic: "large", partition: 0, offset } }] };
      },
    } as unknown as NonNullable<Parameters<typeof createQueuesModule>[0]["gateway"]>;
    const queues = createQueuesModule({ endpoint: "http://gateway", gateway });

    const acquired = await queues.acquire("large", { group: "workers", sessionId: "session" });
    expect(acquired.messages[0]).toMatchObject({ messageId: `large:0:${offset}`, offset });
    await queues.acknowledge("session", [{ messageId: `large:0:${offset}`, ackType: "accept" }]);
  });

  test("live queue payloads distinguish tombstones from empty values", async () => {
    const gateway = {
      async queueAcquire() {
        return {
          sessionId: "session",
          messages: [
            { topic: "queue", partition: 0, offset: 0n, headers: [], deliveryCount: 1 },
            { topic: "queue", partition: 0, offset: 1n, value: new Uint8Array(), headers: [], deliveryCount: 1 },
          ],
        };
      },
    } as unknown as NonNullable<Parameters<typeof createQueuesModule>[0]["gateway"]>;
    const queues = createQueuesModule({ endpoint: "http://gateway", gateway });

    const acquired = await queues.acquire("queue", { group: "workers", sessionId: "session" });
    expect(acquired.messages[0]?.value).toBeUndefined();
    expect(acquired.messages[1]?.value).toEqual(new Uint8Array());
  });

  test("mock acquire validates group and lock duration", async () => {
    const client = createClient({ endpoint: "mock://gateway" });

    await expect(client.queues.acquire("queue", { group: "" })).rejects.toThrow(InvalidArgumentError);
    await expect(client.queues.acquire("queue", { group: "workers", lockDurationMs: 1_000 })).rejects.toMatchObject({
      message: "queue lock_duration_ms must be 30000; per-acquire lock durations are not supported",
    });
    await expect(
      client.queues.acknowledge("session", [{ messageId: "queue:0:0", ackType: "relese" as "release" }]),
    ).rejects.toMatchObject({
      kind: "invalid_argument",
      message: "queue ack_type must be accept, release, or reject",
    });
  });

  test("mock sessions own delivered coordinates and reject unknown reuse", async () => {
    const client = createClient({ endpoint: "mock://gateway" });
    await client.messaging.publish({ topic: "queue", value: encoder.encode("job") });
    const first = await client.queues.acquire("queue", { group: "workers", max: 1 });
    const second = await client.queues.acquire("queue", { group: "workers", max: 1 });

    await expect(
      client.queues.acknowledge(second.sessionId, [{ messageId: first.messages[0]!.messageId, ackType: "accept" }]),
    ).resolves.toEqual({
      results: [
        {
          messageId: first.messages[0]!.messageId,
          error: { kind: "invalid_argument", message: "queue message is not acquired", retriable: false },
        },
      ],
    });
    await expect(
      client.queues.renew(second.sessionId, [{ messageId: first.messages[0]!.messageId }]),
    ).resolves.toEqual({
      results: [
        {
          messageId: first.messages[0]!.messageId,
          error: { kind: "invalid_argument", message: "queue message is not acquired", retriable: false },
        },
      ],
    });
    await expect(
      client.queues.acquire("queue", { group: "workers", sessionId: "missing-session" }),
    ).rejects.toMatchObject({ message: "queue session expired; re-acquire" });
    await expect(
      client.queues.acquire("queue", { group: "other-workers", sessionId: first.sessionId }),
    ).rejects.toMatchObject({ message: "group_id and topics are fixed when a queue session is created" });
  });

  test("mock queue state is independent per group", async () => {
    const client = createClient({ endpoint: "mock://gateway" });
    await client.messaging.publish({ topic: "queue", value: encoder.encode("job") });
    const first = await client.queues.acquire("queue", { group: "first-workers", max: 1 });
    const second = await client.queues.acquire("queue", { group: "second-workers", max: 1 });

    expect(first.messages).toMatchObject([{ messageId: "queue:0:0", deliveryCount: 1 }]);
    expect(second.messages).toMatchObject([{ messageId: "queue:0:0", deliveryCount: 1 }]);
    await expect(
      client.queues.acknowledge(first.sessionId, [{ messageId: "queue:0:0", ackType: "release" }]),
    ).resolves.toEqual({ results: [{ messageId: "queue:0:0", error: null }] });
    await expect(client.queues.renew(second.sessionId, [{ messageId: "queue:0:0" }])).resolves.toEqual({
      results: [{ messageId: "queue:0:0", error: null }],
    });

    const redelivered = await client.queues.acquire("queue", {
      group: "first-workers",
      max: 1,
      sessionId: first.sessionId,
    });
    expect(redelivered.messages).toMatchObject([{ messageId: "queue:0:0", deliveryCount: 2 }]);
    await expect(
      client.queues.acknowledge(first.sessionId, [{ messageId: "queue:0:0", ackType: "accept" }]),
    ).resolves.toEqual({ results: [{ messageId: "queue:0:0", error: null }] });
    await expect(
      client.queues.acknowledge(second.sessionId, [{ messageId: "queue:0:0", ackType: "accept" }]),
    ).resolves.toEqual({ results: [{ messageId: "queue:0:0", error: null }] });
  });

  test("live acknowledge preserves per-entry gateway error classification", async () => {
    const gateway = {
      async queueAcknowledge() {
        return {
          results: [
            { entry: { topic: "queue-ack", partition: 0, offset: 3n }, error: { code: 13, message: "commit failed", retriable: false } },
            { entry: { topic: "queue-ack", partition: 0, offset: 2n }, error: { code: 13, message: "commit retryable", retriable: true } },
            { entry: { topic: "queue-ack", partition: 0, offset: 1n }, error: { code: 9, message: "record is not acquired by this session", retriable: false } },
            { entry: { topic: "queue-ack", partition: 0, offset: 0n }, error: { code: 9, message: "coordinator retry", retriable: true } },
          ],
        };
      },
    } as unknown as NonNullable<Parameters<typeof createQueuesModule>[0]["gateway"]>;
    const queues = createQueuesModule({ endpoint: "http://gateway", gateway });

    await expect(
      queues.acknowledge("queue-session", [
        { messageId: "queue-ack:0:0", ackType: "accept" },
        { messageId: "queue-ack:0:1", ackType: "accept" },
        { messageId: "queue-ack:0:2", ackType: "accept" },
        { messageId: "queue-ack:0:3", ackType: "accept" },
      ]),
    ).resolves.toEqual({
      results: [
        { messageId: "queue-ack:0:3", error: { kind: "server_error", message: "commit failed", retriable: false } },
        { messageId: "queue-ack:0:2", error: { kind: "transport", message: "commit retryable", retriable: true } },
        {
          messageId: "queue-ack:0:1",
          error: { kind: "invalid_argument", message: "record is not acquired by this session", retriable: false },
        },
        { messageId: "queue-ack:0:0", error: { kind: "transport", message: "coordinator retry", retriable: true } },
      ],
    });
  });

  test("live renew preserves per-entry gateway error classification", async () => {
    const gateway = {
      async queueRenew() {
        return {
          results: [
            {
              entry: { topic: "queue-renew", partition: 0, offset: 7n },
              error: { code: 13, message: "renew retryable", retriable: true },
            },
          ],
        };
      },
    } as unknown as NonNullable<Parameters<typeof createQueuesModule>[0]["gateway"]>;
    const queues = createQueuesModule({ endpoint: "http://gateway", gateway });

    await expect(queues.renew("queue-session", [{ messageId: "queue-renew:0:0" }])).resolves.toEqual({
      results: [
        { messageId: "queue-renew:0:7", error: { kind: "transport", message: "renew retryable", retriable: true } },
      ],
    });
  });

  test("live queue results require authoritative response entries", async () => {
    const gateway = {
      async queueAcknowledge() {
        return { results: [{}] };
      },
    } as unknown as NonNullable<Parameters<typeof createQueuesModule>[0]["gateway"]>;
    const queues = createQueuesModule({ endpoint: "http://gateway", gateway });

    await expect(
      queues.acknowledge("queue-session", [{ messageId: "queue-ack:0:0", ackType: "accept" }]),
    ).rejects.toMatchObject({
      kind: "transport",
      message: "queue response result did not include an entry",
    });
  });
});
